### PR TITLE
Basic machinery to put timeline together

### PR DESCRIPTION
**Pull request by derekdowling**
Includes a test suite for testing integrations of:
- timeline
- public posts
- author view
- author posts
- friends

With relevant hooks for all except author posts so far but we probably won't be calling this one.

### CODE REVIEW

File: api/integrations/integrator.py [hunk 1]
```diff
@@ -80,7 +91,7 @@ def send_friend_request(self, author, foreign_author):
             }
         }
         response = self.request(request.post, "friendrequest", data)
-        print vars(response)
+        print vars(response), "response"
```
Comment (derekdowling) [inline api/integrations/integrator.py]: remove

File: api/integrations/integrator.py [hunk 2]
```diff
@@ -62,8 +71,10 @@ def get_author_posts(self, author_id):
         Queries foreign server for /author/:id/posts
         """
         endpoint = "author/%s/posts" % author_id
-        response = self.request(request.get, endpoint).json()["posts"]
-        posts_serializer = PostSerializer(response)
+        response = self.request(request.get, endpoint)
+        print response, "Data"
```
Comment (derekdowling) [inline api/integrations/integrator.py]: remove

**Comment by derekdowling** [comment]
Currently breaking tests.

File: api/tests/test_integrations.py [hunk 3]
```diff
@@ -1,18 +1,94 @@
...
+            return CachedAuthor(
+                host="http://localhost:8001/api/",
+                # host=data["host"],
+                id=data["id"],
+                displayname=data["displayname"]
+            )
+        else:
+            return None
+
+    # def test_get_author_view(self):
```
Comment (derekdowling) [inline api/tests/test_integrations.py:53]: Not currently supported in integrating api.

**Comment by derekdowling** [comment]
There are four tests that are failing right now, two because we still need to work on friends logic, and two to do with integrations. But we should really merge this in because it's way closer to where we need to be.

File: api/integrations/aggregator.py [hunk 4]
```diff
@@ -19,3 +19,16 @@ def get_public_posts():
             node_posts = integrator.get_public_posts()
             posts.extend(node_posts)
         return posts
+
+    @staticmethod
+    # aggregate posts using a list of cached authors
+    def get_posts_for_authors(authors):
+        posts = []
+        for author in authors:
+            print author
```
Comment (xkb) [inline api/integrations/aggregator.py]: remove

File: api/integrations/integrator.py [hunk 5]
```diff
@@ -40,56 +50,81 @@ def get_public_posts(self):
...
-    def get_author_posts(self, author_id):
+    def get_author_posts(self, origin, requestor=None):
         """
         Queries foreign server for /author/:id/posts
         """
-        endpoint = "author/%s/posts" % author_id
-        response = self.request(request.get, endpoint).json()["posts"]
-        posts_serializer = PostSerializer(response)
-        return posts_serializer.data
+        headers = {"UUID": str(requestor)}
```
Comment (xkb) [inline api/integrations/integrator.py:75]: ??

Comment (derekdowling) [inline api/integrations/integrator.py:75]: Don't ask. Requirement for Tamara's teams calls. I'm not going to bother trying to rig up anything more elegant for now.

File: api/integrations/integrator.py [hunk 6]
```diff
@@ -40,56 +50,81 @@ def get_public_posts(self):
...
                 "url": author.url
             }
         }
-        response = self.request(request.post, "friendrequest", data)
-        print vars(response)
-        return response
 
-    def get_author_view(self, author_id):
+        response = self.request(request.post, self.build_url("friendrequest"), data)
+        if response.status_code == 201:
```
Comment (xkb) [inline api/integrations/integrator.py:101]: `201` might not be consistent between teams.

Comment (derekdowling) [inline api/integrations/integrator.py:101]: I'm not going to worry about that for now. It works with them. If it comes to it, I will fix it when we perform other integrations.

Comment (xkb) [inline api/integrations/integrator.py:101]: Okay, mental note then.

File: api/permissions/permissions.py [hunk 7]
```diff
@@ -20,22 +20,13 @@ def isPublic(request, obj):
     return True
 
 def isOnSameHost(request, obj):
-    return obj.author.host == settings.HOST
+    return obj.author.is_local()
```
Comment (xkb) [inline api/permissions/permissions.py:23]: Nice.

File: api/views/content.py [hunk 8]
```diff
@@ -104,14 +105,23 @@ def retrieve(self, request, author_pk=None, pk=None):
...
-            author = get_object_or_404(Author, id=pk)
-            data = AuthorSerializer(author).data
 
-            # append author posts to this object
-            posts = Post.objects.filter(author__id=pk)
-            if posts is not None:
-                posts = PostSerializer(posts, many=True)
-                data["posts"] = posts.data
+            # check if we're querying for a remote author
+            if "HTTP_AUTH_HOST" in request.META:
```
Comment (xkb) [inline api/views/content.py]: `Author-Host` ?

File: api/views/content.py [hunk 9]
```diff
@@ -155,13 +195,20 @@ class PublicPostsViewSet(
...
-        if request.auth is None or request.user.type is not "Node":
-            foreign_posts = Aggregator.get_public_posts()
-            posts.extend(foreign_posts)
+    # dont return public posts of other nodes in node-to-node calls
+    if request.auth is None or request.user.type is not "Node":
+        foreign_posts = Aggregator.get_public_posts()
+        posts.extend(foreign_posts)
 
-        return Response(posts)
+    return posts
```
Comment (xkb) [inline api/views/content.py]: `return {"posts": posts}` ?

File: api/views/content.py (hunk 8 above)
Comment (derekdowling) [inline api/views/content.py]: Okay, updated to match. It performs a regex on an all caps version of the headers provided.

File: api/views/content.py [hunk 10]
```diff
@@ -155,13 +195,20 @@ class PublicPostsViewSet(
     renderer_classes = (PostsJSONRenderer,)
 
     def list(self, request):
-        internal_posts = Post.objects.filter(visibility="PUBLIC")
-        serializer = PostSerializer(internal_posts, many=True)
-        posts = serializer.data
+        return Response(get_public_posts(self, request))
```
Comment (xkb) [inline api/views/content.py]: Probably better to add the posts key in here sorry

File: api/views/content.py (hunk 9 above)
Comment (derekdowling) [inline api/views/content.py]: Smushed the separated function back into the view. I'm just going to pull out the JSONRendered because it's confusing.

**Pull request merged**